python/src/deltachat/events.py

```diff
@@ -111,6 +111,27 @@ def get_info_regex_groups(self, regex, check_error=True):
             if m is not None:
                 return m.groups()
 
+    def wait_for_connectivity(self, connectivity):
+        """Wait for the specified connectivity.
+        This only works reliably if the connectivity doesn't change
+        again too quickly, otherwise we might miss it."""
+        while 1:
+            if self.account.get_connectivity() == connectivity:
+                return
+            self.get_matching("DC_EVENT_CONNECTIVITY_CHANGED")
+
+    def wait_for_connectivity_change(self, previous, expected_next):
+        """Wait until the connectivity changes to `expected_next`.
+        Fails the test if it changes to something else."""
+        while 1:
+            current = self.account.get_connectivity()
```

Comment:
How bad is this? I mean, shouldn't network accesses usually take orders of magnitude longer than executing this code?

The two reasons I did not put the connectivity into the event:

1. there would be a deadlock. See the PR description:

>    `get_connectivity()` does `scheduler.read()`. And `start_io()` does `scheduler.write()`. Therefore, when I tried to do `get_connectivity()` within `ConnectivityStore.set()` for debugging, I got a deadlock (`start_io()` calls `scheduler.start()` which calls `ConnectivityStore.set_not_configured()` which calls `ConnectivityStore.set()`).

If I did `get_connectivity()` within `ConnectivtyStore.set()` in order to put the result into the event, we would have the same problem.

2. get_connectivity() locks quite some mutexes & one RwLock (`scheduler.read()`), and I'm a bit concerned about battery usage.

**How both problems could be solved, at the expense of making things a bit more complicated:**

We could let `ConnectivityStore.set()` lock the scheduler (`scheduler.write()`). The scheduler would then directly contain a mutable map of "folder type -> connectivity", so that `ConnectivityStore.set()` could change the connectivities that way.

Then, as we already have a reference to the scheduler, we could pass it directly to `get_connectivity()` so that we don't have to lock anything within `get_connectivity()`.

And Scheduler.start() could initialize the connectivity without doing any further locks, although this also wouldn't be as straightforward as it currently is.